opm/grid/cpgrid/CpGrid.cpp

```diff
@@ -1362,73 +1362,117 @@ template cpgrid::Entity<3> Dune::createEntity(const CpGrid&, int, bool);
-    // For cells with no parent, we set {-1,-1}. Entries with actual parents will be rewritten.
-    const std::array<int,2>& no_parent = {-1,-1};
-    l1_child_to_parent_cells.resize(data_[1]-> size(0));
-    for (int cell = 0; cell < data_[1]->size(0); ++cell){
-        l1_child_to_parent_cells[cell] = no_parent;
+    const std::vector<int> no_children = {-1};    
+    const std::tuple<int,std::vector<int>> has_no_children = std::make_tuple(-1, no_children);
+    for (int cell = 0; cell < data_[0]-> size(0); ++cell){
+        l0_parent_to_children_cells[cell] = has_no_children;
     }
-    // Rewrite entries for actual parents in level0 and actual children in level1.
-    for (const auto& [parent, children] : parent_to_children_cells) {
-        l0_parent_to_children_cells[parent] = {1, children}; // {level LGR, {child0, child1, ...}}
-        for (const auto& child : children){
-            l1_child_to_parent_cells[child] = {0, parent}; // {level of parent cell, parent cell index in that level}
+    // Get patches corner, face, and cell indices.
+    for (int patch = 0; patch < num_patches; ++patch){
+        const auto& [corners, faces, cells] = (*(this->data_[0])).getPatchGeomIndices(startIJK_vec[patch], endIJK_vec[patch]);
+        all_patch_corners.reserve(corners.size());
+        for (const auto& corner : corners){
+            all_patch_corners.push_back(corner);
         }
-    }
-    // LEVEL 1, definition/declaration of some members:
-    (*data_[1]).level_data_ptr_ = &(this -> data_);
-    (*data_[1]).level_ = 1;
+        all_patch_faces.reserve(faces.size());
+        for (const auto& face : faces){
+            all_patch_faces.push_back(face);
+        }
+        all_patch_cells.reserve(cells.size());
+        for (const auto& cell : cells){
+            all_patch_cells.push_back(cell);
+        }
+        // Build each LGR from the selected patche of cells from level0 (level0 = this->data_[0]).
+        const auto& [level_ptr, boundary_old_to_new_corners, boundary_old_to_new_faces, parent_to_children_faces,
+                     parent_to_children_cells, child_to_parent_faces, child_to_parent_cells]
+            = (*(this-> data_[0])).refinePatch(cells_per_dim_vec[patch], startIJK_vec[patch], endIJK_vec[patch]);
+        // Add each LGR to "data".
+        (this-> data_).push_back(level_ptr);
+        (*data_[patch +1]).level_data_ptr_ = &(this -> data_); // shifted +1 since level0 is coarse grid. Levels are 1,2,..., num_patches.
+        (*data_[patch +1]).level_ = patch +1;
+        //
+        // POPULATING (*data_[0]).parent_to_children_cells_
+        // POPULATING (*data_[patch +1]).child_to_parent_cells_
+        // Extension of container all_patch_child_to_parent_cells, adding {-1,-1} entries for each NO-CHILD cell, in each level/patch.
+        (all_patch_child_to_parent_cells[patch]).resize(child_to_parent_cells.size()); // To consider all cells in the LGR.
```

Comment:
Looking at this line and line I wonder whether we really need the all_patch_child_to_parent_cells container.
We can substitute `all_patch_child_to_parent_cells[patch]` with `level_child_to_parent_cells` an operate directly on that one. That would be defined by
```
auto& level_child_to_parent_cells = this-> data_.back()..child_to_parent_cells_ ;
```

Reply:
For each actual level/LGR, child_to_parent_cells_ has entries {0 (level), parent cell index (in level 0) of that child-cell in that lgr}. For the LeafView, child_to_parent_cells_ has entries {-1,-1} when the cell "was born" in level 0 and has been not involved in any LGR, and {0, parent cell index in level 0} for the cells coming for each LGR. Since the leafview cell index and the "each-level/lgr" index of each cell are different, containers 
(*data_[level]).child_to_parent_cells_
and 
(*data_["leafview"]).child_to_parent_cells_
have different sizes and different entries. 

In the current code, the necessary information to populate child_to_parent_cells_ for the LeafView is deduced from the information created  previously for each level. It is possible to rewrite the code in such a way that the information is taken in the other direction (deduced from LeafView to create the each-level-information).  I would leave it like it is now.